File: android/app/src/main/kotlin/com/cruisemesh/app/identity/backup/BackupService.kt
package com.cruisemesh.app.identity.backup

import android.content.Context
import android.content.pm.ApplicationInfo
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Log
import com.cruisemesh.app.R
import com.cruisemesh.app.identity.IdentityStore
import com.cruisemesh.app.identity.OnboardingStore
import com.cruisemesh.app.identity.ProfilePhotoStore
import com.cruisemesh.app.identity.ProfileStore
import com.cruisemesh.app.friending.FriendsOfFriendsStore
import com.cruisemesh.app.identity.decodeIdentity
import com.cruisemesh.app.identity.encodeIdentity
import com.cruisemesh.app.relay.RelayConfigStore
import com.cruisemesh.app.AppStore
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import uniffi.cruisemesh_core.backupMaxFileBytes
import uniffi.cruisemesh_core.BackupContentOptions
import uniffi.cruisemesh_core.BackupInventory
import uniffi.cruisemesh_core.CoreRestorePlan
import uniffi.cruisemesh_core.inspectRestoredMessageStore
import uniffi.cruisemesh_core.sanitizeRestoredMessageStoreWithOptions

/**
 * Android glue for account backup/restore:
 * gathers the identity + message-store snapshot, seals them with
 * [BackupCrypto], and reads/writes the resulting `.cmbak` through the Storage
 * Access Framework. All calls do KDF + crypto + file I/O and MUST run off the
 * main thread (the callers use a background dispatcher).
 *
 * The message-store snapshot comes from the core's transactionally consistent
 * `MessageStore.backup_to` primitive. The core requires a new absolute path so
 * a snapshot request cannot overwrite another local file.
 */
object BackupService {

    private const val STORE_FILENAME = "cruisemesh.sqlite"
    private const val TAG = "BackupService"

    fun inventory(context: Context, nowMs: Long = System.currentTimeMillis()): BackupInventory =
        AppStore.get(context).backupInventory(nowMs)

    /** Build the encrypted `.cmbak` bytes from the current on-device identity and message store. */
    fun buildBackup(
        context: Context,
        passphrase: CharArray,
        options: BackupContentOptions = defaultContentOptions(),
    ): ByteArray {
        val identity = IdentityStore.load(context)
            ?: throw IllegalStateException("No identity on this device to back up")
        val snapshotFile = java.io.File.createTempFile("cruisemesh-backup-", ".sqlite", context.cacheDir)
        snapshotFile.delete()
        val sqliteBytes = try {
            val report = AppStore.get(context).backupToWithOptions(
                snapshotFile.absolutePath,
                options,
                System.currentTimeMillis(),
            )
            Log.i(
                TAG,
                "Prepared backup snapshot; removed messages=${report.removedMessageCount}, " +
                    "ownPending=${report.removedPendingOwnDeliveryCount}, " +
                    "courier=${report.removedCourierDeliveryCount}, " +
                    "expired=${report.removedExpiredDeliveryCount}, " +
                    "connectionEvents=${report.removedConnectionEventCount}",
            )
            snapshotFile.inputStream().use { input -> readBackupBytes(context, input) }
        } finally {
            snapshotFile.delete()
        }
        val relay = RelayConfigStore.load(context)

        val payload = BackupPayload(
            identity = encodeIdentity(identity),
            sqlite = sqliteBytes,
            srcVersionCode = appVersionCode(context),
            createdAtMs = System.currentTimeMillis(),
            displayName = ProfileStore.loadDisplayName(context),
            ownAvatar = ProfilePhotoStore.loadBackupBytes(context),
            ownAvatarEpoch = ProfileStore.loadOwnAvatarEpoch(context),
            relayUrl = relay?.relayUrl,
            relayToken = relay?.relayToken,
            shareOnline = RelayConfigStore.shareOnline(context),
            friendsOfFriendsEnabled = FriendsOfFriendsStore.isEnabled(context),
        )
        return BackupCrypto.seal(passphrase, payload)
    }

    /**
     * Decrypt and install a backup. Intended for onboarding (fresh install),
     * where the message store has not been opened yet, so overwriting the DB
     * file and re-seeding the identity is safe. Throws a typed
     * [BackupException] on a bad file / wrong passphrase / newer-version backup;
     * on success the caller should restart the app so the identity and store are
     * re-read cleanly.
     */
    fun previewBackup(
        context: Context,
        fileBytes: ByteArray,
        passphrase: CharArray,
    ): BackupPreview {
        val payload = openAndValidatePayload(context, fileBytes, passphrase)
        val inventory = withStagedSqlite(context, payload.sqlite) { staged ->
            inspectRestoredMessageStore(staged.absolutePath, System.currentTimeMillis())
        }
        return BackupPreview(
            inventory = inventory,
            createdAtMs = payload.createdAtMs,
            sourceVersionCode = payload.srcVersionCode,
            // §9's closing paragraph: opening a `.cmbak` on a fresh install is
            // two different intents wearing one word, and the person has to be
            // the one who picks. Read on the decrypt this screen already
            // performs, so choosing does not cost a second passphrase entry.
            plans = BackupCrypto.restorePlans(payload),
        )
    }

    fun restoreBackup(
        context: Context,
        fileBytes: ByteArray,
        passphrase: CharArray,
        options: BackupContentOptions = defaultContentOptions(),
    ) {
        val payload = openAndValidatePayload(context, fileBytes, passphrase)

        val identity = decodeIdentity(payload.identity)
        // Install the message store from a staged file. Do NOT round-trip the
        // sanitized DB through a second heap ByteArray: on a default 256 MiB
        // heap the encrypted backup + decrypted sqlite already fill most of
        // the process, and `File.readBytes()` of an ~88 MiB store OOMs
        // (Pixel 10, 2026-08-07).
        installSanitizedMessageStore(context, payload.sqlite, options)

        // Every preference write below must be durable. The caller restarts by
        // hard-exiting the process, which outruns an `apply()` and used to drop
        // the restored identity, display name and relay endpoint on the floor,
        // leaving a phone with its old messages but a brand-new `user_id`.
        //
        // Re-wrap the restored keys under THIS device's fresh Keystore key.
        IdentityStore.save(context, identity, durable = true)
        payload.displayName?.let { ProfileStore.saveDisplayName(context, it, durable = true) }
        ProfilePhotoStore.restoreBackupBytes(context, payload.ownAvatar)
        ProfileStore.restoreOwnAvatarEpoch(context, payload.ownAvatarEpoch)
        if (payload.relayUrl != null && payload.relayToken != null) {
            RelayConfigStore.save(context, payload.relayUrl, payload.relayToken, durable = true)
        }
        RelayConfigStore.setShareOnline(context, payload.shareOnline, durable = true)
        FriendsOfFriendsStore.restoreEnabled(context, payload.friendsOfFriendsEnabled)
        OnboardingStore.markCompleted(context, durable = true)
    }

    fun defaultContentOptions() = BackupContentOptions(
        includeMessageHistory = true,
        includePendingDeliveriesForOthers = false,
    )

    /** Read a SAF document without ever accumulating more than the core's backup cap. */
    fun readBytes(context: Context, uri: Uri): ByteArray {
        val maxBytes = backupReadLimit()
        val declaredLength = context.contentResolver
            .openAssetFileDescriptor(uri, "r")
            ?.use { it.declaredLength }
        if (declaredLength != null && declaredLength > maxBytes) {
            throw IOException(context.getString(R.string.ui_this_backup_file_is_too_large))
        }
        return context.contentResolver.openInputStream(uri)?.use { input ->
            readBackupBytes(context, input)
        } ?: throw IllegalStateException("Could not open backup file")
    }

    /**
     * The name to show for a picked document. A SAF uri's last path segment is
     * a provider document id, which for a file the app itself wrote is often
     * just a number ("152"); the provider's display-name column is the actual
     * filename. Falls back to the path segment when no provider answers.
     */
    fun displayName(context: Context, uri: Uri): String? {
        val fromProvider = runCatching {
            context.contentResolver
                .query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)
                ?.use { cursor ->
                    val column = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (column >= 0 && cursor.moveToFirst()) cursor.getString(column) else null
                }
        }.getOrNull()
        return fromProvider?.takeIf { it.isNotBlank() }
            ?: uri.lastPathSegment?.substringAfterLast('/')
    }

    /** Write bytes to a SAF document (the destination the user chose to save the backup). */
    fun writeBytes(context: Context, uri: Uri, bytes: ByteArray) {
        // "wt" = write + truncate, so re-saving over an existing file replaces it cleanly.
        context.contentResolver.openOutputStream(uri, "wt")?.use { it.write(bytes) }
            ?: throw IllegalStateException("Could not write backup file")
    }

    /** Suggested filename for a new backup, e.g. `cruisemesh-backup-20260712-1530.cmbak`. */
    fun suggestedFileName(nowMs: Long = System.currentTimeMillis()): String {
        val stamp = java.text.SimpleDateFormat("yyyyMMdd-HHmm", java.util.Locale.US)
            .format(java.util.Date(nowMs))
        return "cruisemesh-backup-$stamp.cmbak"
    }

    private fun appVersionCode(context: Context): Int {
        val info = context.packageManager.getPackageInfo(context.packageName, 0)
        // minSdk is 31, so longVersionCode (added in API 28) is always available.
        return info.longVersionCode.toInt()
    }

    /**
     * True for a locally built app. Read from the installed package rather than
     * `BuildConfig`, which this module does not generate.
     */
    private fun isDebuggableBuild(context: Context): Boolean =
        (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0

    private fun readBackupBytes(context: Context, input: InputStream): ByteArray = try {
        input.readBackupBytes(backupReadLimit())
    } catch (_: BackupFileTooLargeException) {
        throw IOException(context.getString(R.string.ui_this_backup_file_is_too_large))
    }

    private fun backupReadLimit(): Int {
        val limit = backupMaxFileBytes()
        check(limit <= Int.MAX_VALUE.toULong())
        return limit.toInt()
    }

    /**
     * Validate and migrate the SQLite payload away from the installed store,
     * then remove courier and relay-runtime rows before the restored identity
     * can start any transports. This also protects legacy full-DB backups that
     * predate the same scrub in `MessageStore.backupTo`.
     */
    private fun openAndValidatePayload(
        context: Context,
        fileBytes: ByteArray,
        passphrase: CharArray,
    ): BackupPayload {
        val payload = BackupCrypto.open(passphrase, fileBytes)
        val appVersion = appVersionCode(context)
        if (refuseNewerBackup(payload.srcVersionCode, appVersion, isDebuggableBuild(context))) {
            throw BackupException.NewerBackup(payload.srcVersionCode, appVersion)
        }
        if (payload.srcVersionCode > appVersion) {
            Log.w(
                TAG,
                "Reading a backup made by version ${payload.srcVersionCode} on $appVersion; " +
                    "this debuggable build uses a frozen version code.",
            )
        }
        return payload
    }

    /**
     * Sanitize the restored SQLite on a temp path, then place it at the live
     * store path without ever re-materializing the whole DB as a ByteArray.
     */
    private fun installSanitizedMessageStore(
        context: Context,
        sqlite: ByteArray,
        options: BackupContentOptions,
    ) {
        val sqliteFile = context.filesDir.resolve(STORE_FILENAME)
        clearMessageStoreSiblings(context)

        if (sqlite.isEmpty()) {
            sqliteFile.takeIf { it.exists() }?.delete()
            return
        }

        val staged = java.io.File.createTempFile("cruisemesh-restore-", ".sqlite", context.cacheDir)
        var movedToDestination = false
        try {
            staged.writeBytes(sqlite)
            val report = sanitizeRestoredMessageStoreWithOptions(
                staged.absolutePath,
                options,
                System.currentTimeMillis(),
            )
            Log.i(
                TAG,
                "Sanitized restored store; removed messages=${report.removedMessageCount}, " +
                    "ownPending=${report.removedPendingOwnDeliveryCount}, " +
                    "courier=${report.removedCourierDeliveryCount}, " +
                    "expired=${report.removedExpiredDeliveryCount}, " +
                    "connectionEvents=${report.removedConnectionEventCount}",
            )

            // Prefer a same-filesystem rename (zero extra heap). Fall back to a
            // streamed copy if rename is refused (cross-device, or dest open).
            sqliteFile.takeIf { it.exists() }?.delete()
            if (staged.renameTo(sqliteFile)) {
                movedToDestination = true
            } else {
                staged.inputStream().use { input ->
                    sqliteFile.outputStream().use { output -> input.copyTo(output) }
                }
            }
        } finally {
            for (suffix in listOf("-journal", "-wal", "-shm")) {
                java.io.File(staged.path + suffix).takeIf { it.exists() }?.delete()
            }
            if (!movedToDestination) {
                staged.takeIf { it.exists() }?.delete()
            }
        }
    }

    private fun clearMessageStoreSiblings(context: Context) {
        for (suffix in listOf("-journal", "-wal", "-shm")) {
            context.filesDir.resolve(STORE_FILENAME + suffix).takeIf { it.exists() }?.delete()
        }
    }

    private inline fun <T> withStagedSqlite(
        context: Context,
        sqlite: ByteArray,
        block: (java.io.File) -> T,
    ): T {
        val staged = java.io.File.createTempFile("cruisemesh-restore-", ".sqlite", context.cacheDir)
        return try {
            staged.writeBytes(sqlite)
            block(staged)
        } finally {
            for (suffix in listOf("", "-journal", "-wal", "-shm")) {
                staged.resolveSibling(staged.name + suffix).takeIf { it.exists() }?.delete()
            }
        }
    }
}

data class BackupPreview(
    val inventory: BackupInventory,
    val createdAtMs: Long,
    val sourceVersionCode: Int,
    /**
     * §9's "Replace this device" and "Link as new device", in core's order.
     * Empty is never expected — `core_backup_restore_plans` always returns both
     * — but a shell that finds it empty falls back to the old single-meaning
     * restore rather than showing a fork with no branches.
     */
    val plans: List<CoreRestorePlan> = emptyList(),
)

internal class BackupFileTooLargeException : IOException()

internal fun InputStream.readBackupBytes(maxBytes: Int): ByteArray {
    require(maxBytes >= 0)
    val output = ByteArrayOutputStream(minOf(maxBytes, 64 * 1024))
    val buffer = ByteArray(64 * 1024)
    var total = 0
    while (true) {
        val read = read(buffer)
        if (read < 0) break
        if (read == 0) continue
        if (read > maxBytes - total) throw BackupFileTooLargeException()
        output.write(buffer, 0, read)
        total += read
    }
    return output.toByteArray()
}

File: ios/CruiseMesh/Core/BackupService.swift
import Foundation

enum BackupService {
    private static var pendingDatabaseURL: URL {
        AppStore.databaseURL.appendingPathExtension("restore")
    }

    static let defaultContentOptions = BackupContentOptions(
        includeMessageHistory: true,
        includePendingDeliveriesForOthers: false
    )

    static func inventory(nowMs: Int64 = currentTimeMs) throws -> BackupInventory {
        try AppStore.get().backupInventory(nowMs: nowMs)
    }

    static func buildBackup(
        passphrase: String,
        options: BackupContentOptions = defaultContentOptions
    ) throws -> Data {
        guard let identity = IdentityStore.load() else {
            throw BackupServiceError.noIdentity
        }
        let snapshot = FileManager.default.temporaryDirectory
            .appendingPathComponent("cruisemesh-\(UUID().uuidString).sqlite")
        defer { try? FileManager.default.removeItem(at: snapshot) }
        let report = try AppStore.get().backupToWithOptions(
            destination: snapshot.path,
            options: options,
            nowMs: currentTimeMs
        )
        NSLog(
            "Prepared backup snapshot; removed messages=\(report.removedMessageCount) " +
                "ownPending=\(report.removedPendingOwnDeliveryCount) " +
                "courier=\(report.removedCourierDeliveryCount) " +
                "expired=\(report.removedExpiredDeliveryCount) " +
                "connectionEvents=\(report.removedConnectionEventCount)"
        )
        let sqlite = try Data(contentsOf: snapshot)
        let relay = RelayConfigStore.load()
        let payload = CoreBackupPayload(
            identity: encodeIdentityBytes(identity: identity),
            sqlite: sqlite,
            srcVersionCode: appVersionCode,
            createdAtMs: Int64(Date().timeIntervalSince1970 * 1_000),
            displayName: ProfileStore.loadDisplayName(),
            ownAvatar: ProfilePhotoStore.loadBackupBytes(),
            ownAvatarEpoch: ProfileStore.loadOwnAvatarEpoch(),
            relayUrl: relay?.relayUrl,
            relayToken: relay?.relayToken,
            shareOnline: RelayConfigStore.shareOnline(),
            friendsOfFriendsEnabled: FriendsOfFriendsStore.isEnabled()
        )
        return try sealBackup(passphrase: passphrase, payload: payload, iterations: nil)
    }

    static func previewBackup(file: Data, passphrase: String) throws -> BackupPreview {
        let payload = try openAndValidatePayload(file: file, passphrase: passphrase)
        let inventory = try withTemporaryDatabase(payload.sqlite) { staged in
            try inspectRestoredMessageStore(path: staged.path, nowMs: currentTimeMs)
        }
        return BackupPreview(
            inventory: inventory,
            createdAtMs: payload.createdAtMs,
            sourceVersionCode: payload.srcVersionCode,
            // §9's closing paragraph: opening a `.cmbak` on a fresh install is two
            // different intents wearing one word, and the person has to be the one
            // who picks. Read on the decrypt this screen already performs, so
            // choosing does not cost a second passphrase entry.
            //
            // The list is core's, order included — "Link as new device" is
            // deliberately first there, and the surface must not reorder a choice
            // whose ordering is itself the recommendation.
            plans: restorePlans(payload: payload)
        )
    }

    static func stageRestore(
        file: Data,
        passphrase: String,
        options: BackupContentOptions = defaultContentOptions
    ) throws {
        let payload = try openAndValidatePayload(file: file, passphrase: passphrase)
        let identity = try decodeIdentityBytes(bytes: payload.identity)
        try FileManager.default.createDirectory(
            at: pendingDatabaseURL.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try stageSanitizedDatabase(payload.sqlite, options: options)
        IdentityStore.save(identity)
        if let name = payload.displayName { ProfileStore.saveDisplayName(name) }
        ProfilePhotoStore.restoreBackupBytes(payload.ownAvatar)
        ProfileStore.restoreOwnAvatarEpoch(payload.ownAvatarEpoch)
        if let url = payload.relayUrl, let token = payload.relayToken {
            RelayConfigStore.save(relayUrl: url, relayToken: token)
        } else {
            RelayConfigStore.save(relayUrl: "", relayToken: "")
        }
        RelayConfigStore.setShareOnline(payload.shareOnline)
        _ = FriendsOfFriendsStore.setEnabled(payload.friendsOfFriendsEnabled)
        OnboardingStore.markCompleted()
    }

    /// Validate/migrate the payload and remove restored courier/relay runtime
    /// state before placing it at the special path startup knows how to install.
    /// A failed sanitization never leaves an unsafe pending restore behind.
    ///
    /// Mirrors the Android restore path: sanitize on a temp file, then place it
    /// at the pending path without re-materializing the whole DB as `Data`.
    /// Holding the encrypted backup + decrypted sqlite already peaks memory;
    /// `Data(contentsOf:)` of an ~88 MiB store was a third full copy (Android
    /// OOM on Pixel 10 under a 256 MiB heap).
    private static func stageSanitizedDatabase(
        _ sqlite: Data,
        options: BackupContentOptions
    ) throws {
        guard !sqlite.isEmpty else {
            try sqlite.write(to: pendingDatabaseURL, options: .atomic)
            return
        }
        let manager = FileManager.default
        let staged = manager.temporaryDirectory
            .appendingPathComponent("cruisemesh-restore-\(UUID().uuidString).sqlite")
        var movedToDestination = false
        defer {
            for suffix in ["-journal", "-wal", "-shm"] {
                try? manager.removeItem(at: URL(fileURLWithPath: staged.path + suffix))
            }
            if !movedToDestination {
                try? manager.removeItem(at: staged)
            }
        }
        try sqlite.write(to: staged, options: .atomic)
        let report = try sanitizeRestoredMessageStoreWithOptions(
            path: staged.path,
            options: options,
            nowMs: currentTimeMs
        )
        NSLog(
            "Sanitized restored store; removed messages=\(report.removedMessageCount) " +
                "ownPending=\(report.removedPendingOwnDeliveryCount) " +
                "courier=\(report.removedCourierDeliveryCount) " +
                "expired=\(report.removedExpiredDeliveryCount) " +
                "connectionEvents=\(report.removedConnectionEventCount)"
        )
        try relocateStagedDatabase(from: staged, to: pendingDatabaseURL, fileManager: manager)
        movedToDestination = true
    }

    /// Places a staged SQLite file at the pending restore path without loading
    /// the whole DB into a third `Data` buffer. Prefer rename; fall back to
    /// file-to-file copy if move is refused. On success the staged URL is
    /// always consumed (removed after a copy fallback).
    ///
    /// Internal (not private) so unit tests can drive the same path restore
    /// uses after sanitize — the step that used to OOM on Android via
    /// `Data(contentsOf:)`.
    static func relocateStagedDatabase(
        from staged: URL,
        to destination: URL,
        fileManager: FileManager = .default
    ) throws {
        if fileManager.fileExists(atPath: destination.path) {
            try fileManager.removeItem(at: destination)
        }
        do {
            try fileManager.moveItem(at: staged, to: destination)
        } catch {
            try fileManager.copyItem(at: staged, to: destination)
            try? fileManager.removeItem(at: staged)
        }
    }

    private static func openAndValidatePayload(
        file: Data,
        passphrase: String
    ) throws -> CoreBackupPayload {
        let payload = try openBackup(passphrase: passphrase, file: file)
        guard payload.srcVersionCode <= appVersionCode else {
            throw BackupServiceError.newerBackup(payload.srcVersionCode)
        }
        return payload
    }

    private static func withTemporaryDatabase<T>(
        _ sqlite: Data,
        operation: (URL) throws -> T
    ) throws -> T {
        let manager = FileManager.default
        let staged = manager.temporaryDirectory
            .appendingPathComponent("cruisemesh-restore-\(UUID().uuidString).sqlite")
        defer {
            for suffix in ["", "-journal", "-wal", "-shm"] {
                try? manager.removeItem(at: URL(fileURLWithPath: staged.path + suffix))
            }
        }
        try sqlite.write(to: staged, options: .atomic)
        return try operation(staged)
    }

    /// Read a selected backup incrementally so a malicious file provider cannot
    /// make the app accumulate an unbounded `Data` value before core validation.
    static func readBackupFile(
        at url: URL,
        maxBytes: Int = Int(backupMaxFileBytes())
    ) throws -> Data {
        precondition(maxBytes >= 0)
        let values = try url.resourceValues(forKeys: [.fileSizeKey])
        if let fileSize = values.fileSize, fileSize > maxBytes {
            throw BackupServiceError.fileTooLarge
        }
        guard let input = InputStream(url: url) else {
            throw CocoaError(.fileReadUnknown)
        }
        input.open()
        defer { input.close() }

        let chunkSize = 64 * 1024
        var buffer = [UInt8](repeating: 0, count: chunkSize)
        var result = Data()
        result.reserveCapacity(min(maxBytes, chunkSize))
        while true {
            let count = buffer.withUnsafeMutableBytes { rawBuffer -> Int in
                guard let baseAddress = rawBuffer.baseAddress else { return 0 }
                return input.read(
                    baseAddress.assumingMemoryBound(to: UInt8.self),
                    maxLength: chunkSize
                )
            }
            if count < 0 {
                throw input.streamError ?? CocoaError(.fileReadUnknown)
            }
            if count == 0 { break }
            guard count <= maxBytes - result.count else {
                throw BackupServiceError.fileTooLarge
            }
            result.append(contentsOf: buffer.prefix(count))
        }
        return result
    }

    /// Called before `AppStore` or `MeshController` is initialized, so the
    /// existing SQLite connection can never observe a file replacement.
    static func installPendingRestoreIfNeeded() throws {
        guard FileManager.default.fileExists(atPath: pendingDatabaseURL.path) else { return }
        let manager = FileManager.default
        let destination = AppStore.databaseURL
        try manager.createDirectory(at: destination.deletingLastPathComponent(), withIntermediateDirectories: true)
        for suffix in ["-journal", "-wal", "-shm"] {
            let sibling = URL(fileURLWithPath: destination.path + suffix)
            if manager.fileExists(atPath: sibling.path) { try manager.removeItem(at: sibling) }
        }
        if manager.fileExists(atPath: destination.path) { try manager.removeItem(at: destination) }
        // Empty-store restore writes a zero-length pending file. Check size
        // only — do not load the pending DB into memory just to test emptiness.
        let size = try manager.attributesOfItem(atPath: pendingDatabaseURL.path)[.size] as? NSNumber
        if let size, size.intValue > 0 {
            try manager.moveItem(at: pendingDatabaseURL, to: destination)
        } else {
            try manager.removeItem(at: pendingDatabaseURL)
        }
    }

    static var suggestedFileName: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyyMMdd-HHmm"
        return "cruisemesh-backup-\(formatter.string(from: Date())).cmbak"
    }

    private static var appVersionCode: Int32 {
        Int32(Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "0") ?? 0
    }

    private static var currentTimeMs: Int64 {
        Int64(Date().timeIntervalSince1970 * 1_000)
    }
}

struct BackupPreview: Equatable {
    let inventory: BackupInventory
    let createdAtMs: Int64
    let sourceVersionCode: Int32
    /// §9's "Replace this device" and "Link as new device", in core's order.
    /// Empty is never expected — `core_backup_restore_plans` always returns both
    /// — but a shell that finds it empty falls back to the old single-meaning
    /// restore rather than showing a fork with no branches.
    var plans: [CoreRestorePlan] = []
}

/// §9's two intents, or the one that keeps a restore working if core cannot
/// produce them.
///
/// `core_backup_restore_plans` always returns both, so the fallback is not an
/// expected state. It exists because the alternative — an empty list — silently
/// removes every branch from the fork, leaving a person who opened a backup with
/// a screen that offers nothing and a Restore button that never enables. Seeding
/// the replace intent degrades to exactly the behaviour restore had before the
/// fork existed, which is the honest floor: it is the branch that finishes here,
/// and the one whose hazard the copy already warns about.
private func restorePlans(payload: CoreBackupPayload) -> [CoreRestorePlan] {
    if let plans = try? coreBackupRestorePlans(payload: payload), !plans.isEmpty {
        return plans
    }
    // The person id comes out of the identity block the backup already carries;
    // with neither that nor core's plans there is genuinely nothing to offer.
    guard let identity = try? decodeIdentityBytes(bytes: payload.identity) else { return [] }
    return [
        CoreRestorePlan(
            intent: .replaceThisDevice,
            personId: identity.userId,
            restoresStoredHistory: true,
            keepsExistingDeviceIdentity: true,
            mintsNewDeviceKey: false,
            routesToLinkCeremony: false,
            carriesRecoveryMaterial: false,
            cloneHazardIfSourceIsLive: true
        )
    ]
}

enum BackupServiceError: LocalizedError, Equatable {
    case noIdentity
    case newerBackup(Int32)
    case fileTooLarge

    var errorDescription: String? {
        // One source of copy: the same sentences the backup screens show.
        switch self {
        case .noIdentity: return BackupFailureReason.noAccountToBackUp.text
        case .newerBackup: return BackupFailureReason.newerVersion.text
        case .fileTooLarge: return BackupFailureReason.tooLarge.text
        }
    }
}
